### PR TITLE
add PurlListener to watch for recent changes in published PURLs

### LINKED ISSUE #66: Add listener code to purl-fetcher
- [ ] implement listener process using gem (in #104)
- [ ] integrate listener process management with capistrano (in #104)
- [ ] develop entry point for listener to make index changes (in #104)
- [ ] document administrative use (where should this doc live?)
- [ ] implement changes to the `publish` robot (in #103)

Also see #61

### LINKED ISSUE #106: Fix Honeybadger exception logging on listener:restart
See https://app.honeybadger.io/projects/48916/faults/30715349

See #104

### PR DESCRIPTION
**Pull request by drh-stanford**
Paired with @eefahy on this PR.

This PR fixes #66 and fixes #106 by adding `PurlListener` to support incremental indexing. The listener is a daemon process and includes the `listener:start/stop/restart/status` tasks. A `cap deploy` and `cap deploy:restart` will restart the listener daemon.

See https://github.com/sul-dlss/purl-fetcher/issues/66#issuecomment-239324781 for a discussion of the basic algorithm.

**NOTE**: That this PR requires puppet changes before deployment. Right now only the `-dev` environment has those changes.

### CODE REVIEW

File: Capfile [hunk 1]
```diff
@@ -26,7 +26,8 @@ require 'dlss/capistrano'
 require 'capistrano/rvm'
 require 'whenever/capistrano'
 require 'capistrano/honeybadger'
-require 'capistrano/rails/migrations'
+require 'capistrano/rails'
```
Comment (mejackreed) [inline Capfile]: Why the add back of assets precompile?

Comment (drh-stanford) [inline Capfile]: I think Erin and I were experimenting with getting the `listener:restart` task run by cap deploy and this snuck in here. I'll restore it.

File: config/schedule.rb [hunk 2]
```diff
@@ -2,10 +2,6 @@
 
 set :output, '/dev/null'
 
-every 1.day do
```
Comment (mejackreed) [inline config/schedule.rb]: Why do we remove this now? I wonder could we make `find:all` idempotent and then still run that daily?

File: lib/purl_listener.rb [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+  attr_reader :event_handler, :logger, :path, :pid_file
+
+  ##
+  # @param [Pathname] `path` -- the single "dropbox" directory
+  # @param [Pathname] `pid_file` -- the location of the pid file
+  # @param [Logger] `logger` -- place to log info/warn/errors
+  # @raise [ArgumentError] -- requires path to be a directory
+  def initialize(path = Pathname(PurlFetcher::Application.config.app_config['listener_path']),
+                 pid_file = Pathname(PurlFetcher::Application.config.app_config['listener_pid_file']),
+                 logger = IndexingLogger)
```
Comment (mejackreed) [inline lib/purl_listener.rb]: Is there a reason we use `IndexingLogger` here? Why not just the Rails logger?

**Comment by mejackreed** [comment]
Looks good.. just a couple super minor questions/comments. Nothing blocking this though

File: config/schedule.rb (hunk 2 above)
Comment (drh-stanford) [inline config/schedule.rb]: Running it daily incurs the cost of the `find` command, but I don't think that's a big deal. The real question is whether the code handles unchanged `public` files correctly. So, I think it'd be another issue to fire off to at least verify that it does.

File: lib/purl_listener.rb (hunk 3 above)
Comment (drh-stanford) [inline lib/purl_listener.rb]: Yes, `IndexingLogger` goes to the `log/indexing.log` file where as Rails logger goes to `log/production.log`. I think the idea is to keep all the "indexing" activity distinct from the API activity.

Comment (mejackreed) [inline lib/purl_listener.rb]: gotcha.. thx

**Pull request merged**